lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -77,16 +77,43 @@ sub list {
         };
         push @list, $data;
     }
-    @list = sort { $b->{job}->t_started <=> $a->{job}->t_started || $b->{job}->id <=> $a->{job}->id } @list;
+    @list = sort {
+        if ($b->{job} && $a->{job}) {
+            $b->{job}->t_started <=> $a->{job}->t_started || $b->{job}->id <=> $a->{job}->id;
+        }
+        elsif ($b->{job}) {
+            1;
+        }
+        elsif ($a->{job}) {
+            -1;
+        }
+        else {
+            0;
+        }
+    } @list;
     $self->stash(running => \@list);
 
     my @scheduled = $self->db->resultset("Jobs")->complex_query(
-        state   => 'scheduled',
+        state   => [OpenQA::Schema::Result::Jobs::PRE_EXECUTION_STATES],
         match   => $match,
         groupid => $groupid,
         assetid => $assetid
     )->all;
-    @scheduled = sort { $b->t_created <=> $a->t_created || $b->id <=> $a->id } @scheduled;
+    # @scheduled = sort { $b->t_created || 'Inf' <=> $a->t_created || 'Inf' || $b->id <=> $a->id } @scheduled;
+    @scheduled = sort {
+        if ($b->{job} && $a->{job}) {
+            $b->{job}->t_created <=> $a->{job}->t_created || $b->{job}->id <=> $a->{job}->id;
+        }
+        elsif ($b->{job}) {
+            1;
+        }
+        elsif ($a->{job}) {
+            -1;
+        }
+        else {
+            0;
+        }
+    } @scheduled;
```

Comment:
Please [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself). You are calling the same code multiple times, please [extract method](http://refactoring.com/catalog/extractMethod.html).

Reply:
It is actually not the same code. I guess it could be extracted so that we end up with something like `$b->$time_key <=> $a->$time_key || $b->id <=> $a->id`